File: native/iOS/test/www/NativeControls2.js
/*
//  This code is adapted from the work of:
//  Created by Michael Nachbaur on 13/04/09.
//  Copyright 2009 Decaf Ninja Software. All rights reserved.
//  MIT licensed
*/

/**
 * This class exposes mobile phone interface controls to JavaScript, such as
 * native tab and tool bars, etc.
 * @constructor
 */
function NativeControls2() {
    this.tabBarTag = 0;
    this.toolBarTag = 0;
    this.tabBarCallbacks = {};
    this.toolBarCallbacks = {};
}

NativeControls2.prototype._createCustomEventName = function(type, name) {
	// name is the name of the tool/tab bar item, type is "click" for e.g
	return 'command-' + name + '-' + type;
};


/**
 * Create a native tab bar that can have tab buttons added to it which can respond to events.
 */
NativeControls2.prototype.createTabBar = function() {
    PhoneGap.exec("NativeControls2.createTabBar");
};

/**
 * Remove a native tab bar
 */
NativeControls2.prototype.removeTabBar = function() {
    PhoneGap.exec("NativeControls2.removeTabBar");
};

/**
 * Show a tab bar.  The tab bar has to be created first.
 * @param {Object} [options] Options indicating how the tab bar should be shown:
 * - \c height integer indicating the height of the tab bar (default: \c 49)
 * - \c position specifies whether the tab bar will be placed at the \c top or \c bottom of the screen (default: \c bottom)
 */
NativeControls2.prototype.showTabBar = function(animate) {
    if (animate == undefined || animate == null) {
        animate = true;
	}
    PhoneGap.exec("NativeControls2.showTabBar", { animate: animate });
};

/**
 * Hide a tab bar.  The tab bar has to be created first.
 */
NativeControls2.prototype.hideTabBar = function(animate) {
    if (animate == undefined || animate == null) {
        animate = true;
	}
    PhoneGap.exec("NativeControls2.hideTabBar", { animate: animate });
};

/**
 * Create a new tab bar item for use on a previously created tab bar.  Use ::showTabBarItems to show the new item on the tab bar.
 *
 * If the supplied image name is one of the labels listed below, then this method will construct a tab button
 * using the standard system buttons.  Note that if you use one of the system images, that the \c title you supply will be ignored.
 *
 * <b>Tab Buttons</b>
 *   - tabButton:More
 *   - tabButton:Favorites
 *   - tabButton:Featured
 *   - tabButton:TopRated
 *   - tabButton:Recents
 *   - tabButton:Contacts
 *   - tabButton:History
 *   - tabButton:Bookmarks
 *   - tabButton:Search
 *   - tabButton:Downloads
 *   - tabButton:MostRecent
 *   - tabButton:MostViewed
 * @param {String} name internal name to refer to this tab by
 * @param {String} [title] title text to show on the tab, or null if no text should be shown
 * @param {String} [image] image filename or internal identifier to show, or null if now image should be shown
 * @param {Object} [options] Options for customizing the individual tab item
 *  - \c badge value to display in the optional circular badge on the item; if null or unspecified, the badge will be hidden
 */
NativeControls2.prototype.createTabBarItem = function(name, label, image, enabled, options) {
    
	var tag = this.tabBarTag++;
    if (options && 'onSelect' in options && typeof(options['onSelect']) == 'function') {
        this.tabBarCallbacks[tag] = {'onSelect':options.onSelect,'name':name};
    } else {
        this.tabBarCallbacks[tag] = { 'name':name };
	}
	
	if (enabled == null) {
		enabled = true;
	}
	
    PhoneGap.exec("NativeControls2.createTabBarItem", name, (label || ""), image, enabled, tag, options);
};

/**
 * Update an existing tab bar item to change its badge value.
 * @param {String} name internal name used to represent this item when it was created
 * @param {Object} options Options for customizing the individual tab item
 *  - \c badge value to display in the optional circular badge on the item; if null or unspecified, the badge will be hidden
 */
NativeControls2.prototype.updateTabBarItem = function(name, options) {
    if (!options) options = {};
    PhoneGap.exec("NativeControls2.updateTabBarItem", name, options);
};

/**
 * Remove an existing tab bar item.
 * @param {String} name internal name used to represent this item when it was created
 * @param {Object} options Options for customizing the individual tab item
 */
NativeControls2.prototype.removeTabBarItem = function(name, options) {
    if (!options) options = {};
    PhoneGap.exec("NativeControls2.removeTabBarItem", name, options);
};

/**
 * Enable/disable existing tab bar item.
 * @param {String} name internal name used to represent this item when it was created
 * @param {Object} options Options for customizing the individual tab item
 */
NativeControls2.prototype.enableTabBarItem = function(name, enabled) {
    PhoneGap.exec("NativeControls2.enableTabBarItem", name, (enabled || false));
};


/**
 * Manually select an individual tab bar item, or nil for deselecting a currently selected tab bar item.
 * @param {String} tabName the name of the tab to select, or null if all tabs should be deselected
 * @see createTabBarItem
 * @see showTabBarItems
 */
NativeControls2.prototype.selectTabBarItem = function(tab) {
    PhoneGap.exec("NativeControls2.selectTabBarItem", tab);
};

/**
 * Function called when a tab bar item has been selected.
 * @param {Number} tag the tag number for the item that has been selected
 */
NativeControls2.prototype.tabBarItemSelected = function(tag) 
{
	try {

    if (this.tabBarCallbacks[tag] && typeof(this.tabBarCallbacks[tag].onSelect) == 'function') {
        this.tabBarCallbacks[tag].onSelect(this.tabBarCallbacks[tag].name);
	}
		
	
	var name = this.tabBarCallbacks[tag].name;
	// also fire custom event
	if (name) {
		var eventName = this._createCustomEventName('click', this.tabBarCallbacks[tag].name);
		var e = document.createEvent('Events');
		e.initEvent(eventName);
		console.log("Firing event: " + eventName);
		window.dispatchEvent(e);
	}
		
	} catch (e) {
		alert(e);
	}
};

// /////////////////////////////////////////////////////////////////////////////////

/**
 * Remove a toolbar.
 */
NativeControls2.prototype.removeToolBar = function() 
{
    PhoneGap.exec("NativeControls2.removeToolBar");
};

/**
 * Create a toolbar.
 */
NativeControls2.prototype.createToolBar = function() 
{
    PhoneGap.exec("NativeControls2.createToolBar");
};

/**
 * Create a new toolbar item for use on a previously created toolbar.
 *
 * If the supplied image name is one of the labels listed below, then this method will construct a toolbar button
 * using the standard system buttons.  Note that if you use one of the system images, that the \c title you supply will be ignored.
 *
 * <b>Toolbar Buttons</b>
 *   - toolbarButton:Done      
 *   - toolbarButton:Cancel   
 *   - toolbarButton:Edit
 *   - toolbarButton:Save
 *   - toolbarButton:Add
 *   - toolbarButton:FlexibleSpace 
 *   - toolbarButton:FixedSpace
 *   - toolbarButton:Compose
 *   - toolbarButton:Reply
 *   - toolbarButton:Action
 *   - toolbarButton:Organize
 *   - toolbarButton:Bookmarks
 *   - toolbarButton:Search
 *   - toolbarButton:Refresh
 *   - toolbarButton:Stop
 *   - toolbarButton:Camera
 *   - toolbarButton:Trash
 *   - toolbarButton:Play
 *   - toolbarButton:Pause
 *   - toolbarButton:Rewind
 *   - toolbarButton:FastForward
 *   - toolbarButton:Undo
 *   - toolbarButton:Redo
 *
 * @param {String} name internal name to refer to this toolbar item by
 * @param {String} [title] title text to show on the toolbar item, or null if no text should be shown
 * @param {String} [image] image filename or internal identifier to show, or null if now image should be shown
 * @param {Object} [options] Options for customizing the individual toolbar item
 */
NativeControls2.prototype.createToolBarItem = function(name, label, image, enabled, options) {
    
	var tag = this.toolBarTag++;
    if (options && 'onSelect' in options && typeof(options['onSelect']) == 'function') {
        this.toolBarCallbacks[tag] = {'onSelect':options.onSelect,'name':name };
    } else {
        this.toolBarCallbacks[tag] = { 'name':name };
	}
	
	if (enabled == null) {
		enabled = true;
	}
	
    PhoneGap.exec("NativeControls2.createToolBarItem", name, (label || ""), image, enabled, tag, options);
};

/**
 * Remove an existing toolbar item.
 * @param {String} name internal name used to represent this item when it was created
 * @param {Object} options Options for customizing the individual tab item
 */
NativeControls2.prototype.removeToolBarItem = function(name, options) {
    if (!options) options = {};
    PhoneGap.exec("NativeControls2.removeToolBarItem", name, options);
};

/**
 * Enable/disable existing toolbar item.
 * @param {String} name internal name used to represent this item when it was created
 * @param {Object} options Options for customizing the individual tab item
 */
NativeControls2.prototype.enableToolBarItem = function(name, enable) {
    PhoneGap.exec("NativeControls2.enableToolBarItem", name, enable || false);
};

/**
 * Function called when a toolbar item has been selected.
 * @param {Number} tag the tag number for the item that has been selected
 */
NativeControls2.prototype.toolBarItemSelected = function(tag) 
{
    if (this.toolBarCallbacks[tag] && typeof(this.toolBarCallbacks[tag].onSelect) == 'function') {
        this.toolBarCallbacks[tag].onSelect(this.toolBarCallbacks[tag].name);
	}
	
	var name = this.toolBarCallbacks[tag].name;
	// also fire custom event
	if (name) {
		var eventName = this._createCustomEventName('click', this.toolBarCallbacks[tag].name);
		var e = document.createEvent('Events');
		e.initEvent(eventName);
		window.dispatchEvent(e);
	}
};

NativeControls2.prototype.showToolBar = function(animate) {
    if (animate == undefined || animate == null) {
        animate = true;
	}
    PhoneGap.exec("NativeControls2.showToolBar", { animate: animate });
};

NativeControls2.prototype.hideToolBar = function(animate) {
    if (animate == undefined || animate == null) {
        animate = true;
	}
    PhoneGap.exec("NativeControls2.hideToolBar", { animate: animate });
};

NativeControls2.install = function()
{
	if(!window.plugins)	{
		window.plugins = {};
	}
	
	if (!window.plugins.nativeControls) {
		window.plugins.nativeControls = new NativeControls2();
	}
};

PhoneGap.addConstructor(NativeControls2.install);

File: native/ios/NativeControls/NativeControls2.m
//
//  NativeControls.h
//  
//
//  Created by Jesse MacFadyen on 10-02-03.
//  Improved by Shazron Abdullah 11-06-03
//  MIT Licensed

//  Originally this code was developed my Michael Nachbaur
//  Formerly -> PhoneGap :: UIControls.h
//  Created by Michael Nachbaur on 13/04/09.
//  Copyright 2009 Decaf Ninja Software. All rights reserved.

#import <PhoneGap/PhoneGapDelegate.h>
#import "NativeControls2.h"

#import <QuartzCore/QuartzCore.h>

@implementation NativeControls2

@synthesize tabBar, toolBar, tabBarItems, toolBarItems;
@synthesize lastTabBarPosition, lastToolBarPosition;

-(PGPlugin*) initWithWebView:(UIWebView*)theWebView
{
    self = (NativeControls2*)[super initWithWebView:theWebView];
    if (self) 
	{
        self.tabBarItems = [[NSMutableDictionary alloc] initWithCapacity:5];
        self.toolBarItems = [[NSMutableDictionary alloc] initWithCapacity:5];
    }
    return self;
}

- (void) dealloc
{
	self.tabBar = nil;
	self.toolBar = nil;
	self.toolBarItems = nil;
	self.tabBarItems = nil;
	
    [super dealloc];
}

#pragma mark -
#pragma mark TabBar

- (void) removeTabBar:(NSArray*)arguments withDict:(NSDictionary*)options
{
	if (!self.tabBar) {
		return;
	}
	
	[self.webView pg_removeSiblingView:self.tabBar withAnimation:NO];
	[self.webView pg_relayout:NO];
	
	self.tabBar = nil;
}


- (void) createTabBar:(NSArray*)arguments withDict:(NSDictionary*)options
{
	if (self.tabBar) {
		return;
	}
	
    self.tabBar = [UITabBar new];
    self.tabBar.delegate = self;
    self.tabBar.multipleTouchEnabled   = NO;
    self.tabBar.autoresizesSubviews    = YES;
    self.tabBar.hidden                 = NO;
    self.tabBar.userInteractionEnabled = YES;
	self.tabBar.opaque = YES;
    [self.tabBar sizeToFit];
	
	self.webView.superview.autoresizesSubviews = YES;
	
	BOOL atBottom = YES;
	
    if (options) {
        atBottom = [[options objectForKey:@"position"] isEqualToString:@"bottom"];
    }
	
	self.lastTabBarPosition = atBottom?PGLayoutPositionBottom:PGLayoutPositionTop;
	[self.webView pg_addSiblingView:self.tabBar withPosition:self.lastTabBarPosition withAnimation:NO];
}

/**
 * Show the tab bar after its been created.
 */
- (void) showTabBar:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.tabBar) {
        [self createTabBar:nil withDict:nil];
	}
    
	if (![self.webView pg_hasSiblingView:self.tabBar]) {
		[self.webView pg_addSiblingView:self.tabBar withPosition:self.lastTabBarPosition withAnimation:NO];
	}
}

/**
 * Hide the tab bar
 */
- (void) hideTabBar:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.tabBar) {
        return;
	}
	
	if ([self.webView pg_hasSiblingView:self.tabBar]) {
		[self.webView pg_removeSiblingView:self.tabBar withAnimation:NO];
		[self.webView pg_relayout:NO];
	}
}

/**
 * Create a new tab bar item for use on a previously created tab bar.
 *
 * If the supplied image name is one of the labels listed below, then this method will construct a tab button
 * using the standard system buttons.  Note that if you use one of the system images, that the \c title you supply will be ignored.
 * - <b>Tab Buttons</b>
 *   - tabButton:More
 *   - tabButton:Favorites
 *   - tabButton:Featured
 *   - tabButton:TopRated
 *   - tabButton:Recents
 *   - tabButton:Contacts
 *   - tabButton:History
 *   - tabButton:Bookmarks
 *   - tabButton:Search
 *   - tabButton:Downloads
 *   - tabButton:MostRecent
 *   - tabButton:MostViewed
 * @brief create a tab bar item
 * @param arguments Parameters used to create the tab bar
 *  -# \c name internal name to refer to this tab by
 *  -# \c title title text to show on the tab, or null if no text should be shown
 *  -# \c image image filename or internal identifier to show, or null if now image should be shown
 *  -# \c tag unique number to be used as an internal reference to this button
 * @param options Options for customizing the individual tab item
 *  - \c badge value to display in the optional circular badge on the item; if nil or unspecified, the badge will be hidden
 */
- (void) createTabBarItem:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.tabBar) {
        [self createTabBar:nil withDict:nil];
	}

    NSString* name      = [arguments objectAtIndex:0];
    NSString* title     = [arguments objectAtIndex:1];
    NSString* imageName = [arguments objectAtIndex:2];
    BOOL enable         = [[arguments objectAtIndex:3] boolValue];
    int tag             = [[arguments objectAtIndex:4] intValue];

    UITabBarItem *item = nil;    
    if ([imageName length] > 0) 
	{
        UITabBarSystemItem systemItem = -1;
        if ([imageName isEqualToString:@"tabButton:More"])       systemItem = UITabBarSystemItemMore;
        if ([imageName isEqualToString:@"tabButton:Favorites"])  systemItem = UITabBarSystemItemFavorites;
        if ([imageName isEqualToString:@"tabButton:Featured"])   systemItem = UITabBarSystemItemFeatured;
        if ([imageName isEqualToString:@"tabButton:TopRated"])   systemItem = UITabBarSystemItemTopRated;
        if ([imageName isEqualToString:@"tabButton:Recents"])    systemItem = UITabBarSystemItemRecents;
        if ([imageName isEqualToString:@"tabButton:Contacts"])   systemItem = UITabBarSystemItemContacts;
        if ([imageName isEqualToString:@"tabButton:History"])    systemItem = UITabBarSystemItemHistory;
        if ([imageName isEqualToString:@"tabButton:Bookmarks"])  systemItem = UITabBarSystemItemBookmarks;
        if ([imageName isEqualToString:@"tabButton:Search"])     systemItem = UITabBarSystemItemSearch;
        if ([imageName isEqualToString:@"tabButton:Downloads"])  systemItem = UITabBarSystemItemDownloads;
        if ([imageName isEqualToString:@"tabButton:MostRecent"]) systemItem = UITabBarSystemItemMostRecent;
        if ([imageName isEqualToString:@"tabButton:MostViewed"]) systemItem = UITabBarSystemItemMostViewed;
		
        if (systemItem != -1) {
            item = [[UITabBarItem alloc] initWithTabBarSystemItem:systemItem tag:tag];
		}
    }
    
    if (item == nil) {
        NSLog(@"Creating with custom image and title");
		UIImage* image = [UIImage imageNamed:imageName];
		if (!image) {
			NSString* imagePath = [[PhoneGapDelegate class] pathForResource:imageName];
			image = [UIImage imageWithContentsOfFile:imagePath];
		}
		
        item = [[UITabBarItem alloc] initWithTitle:title image:image tag:tag];
    }

    if ([options objectForKey:@"badge"]) {
        item.badgeValue = [options objectForKey:@"badge"];
	}
    
	item.enabled = enable;
	
    BOOL animateItems = YES;
    if ([options objectForKey:@"animate"]) {
        animateItems = [(NSString*)[options objectForKey:@"animate"] boolValue];
	}
	
    NSMutableArray* items = [[self.tabBar items] mutableCopy];
	if (!items) {
		items = [[NSMutableArray alloc] initWithCapacity:1];
	}
	[items addObject:item];
	
    [self.tabBar setItems:items animated:animateItems];
	[items release];
	
	[self.tabBarItems setObject:item forKey:name];
	[item release];
}


/**
 * Update an existing tab bar item to change its title/badge value.
 */
- (void) updateTabBarItem:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.tabBar) {
        [self createTabBar:nil withDict:nil];
	}

    NSString* name = [arguments objectAtIndex:0];
    UITabBarItem* item = [self.tabBarItems objectForKey:name];
    if (item) {
		item.title = [options objectForKey:@"title"];
        item.badgeValue = [options objectForKey:@"badge"];
	}
}

/**
 * Remove an existing tab bar item
 * @param arguments Parameters used to identify the tab bar item to update
 *  -# \c name internal name used to represent this item when it was created
 */
- (void) removeTabBarItem:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.tabBar) {
        [self createTabBar:nil withDict:nil];
	}
	
    NSString* name = [arguments objectAtIndex:0];
    UITabBarItem* item = [self.tabBarItems objectForKey:name];
	
	NSMutableArray* items = [self.tabBar.items mutableCopy];
	NSUInteger index = [items indexOfObject:item];
	if (index != NSNotFound) {
		[items removeObjectAtIndex:index];
		[self.tabBar setItems:items animated:YES];
	}
	
	[items release];
}

- (void) enableTabBarItem:(NSArray*)arguments withDict:(NSDictionary*)options
{
    NSString* name      = [arguments objectAtIndex:0];
    BOOL enable         = [[arguments objectAtIndex:1] boolValue];
	
    UITabBarItem* item = [self.tabBarItems objectForKey:name];
	
	NSUInteger index = [self.tabBar.items indexOfObject:item];
	if (index != NSNotFound) {
		item.enabled = enable;
	}
}

/**
 * Manually select an individual tab bar item, or nil for deselecting a currently selected tab bar item.
 * @brief manually select a tab bar item
 * @param arguments the name of the tab bar item to select
 * @see createTabBarItem
 * @see showTabBarItems
 */
- (void) selectTabBarItem:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.tabBar) {
        [self createTabBar:nil withDict:nil];
	}

    NSString* itemName = [arguments objectAtIndex:0];
    UITabBarItem* item = [self.tabBarItems objectForKey:itemName];
    self.tabBar.selectedItem = item;
}

- (void)tabBar:(UITabBar*)tabBar didSelectItem:(UITabBarItem *)item
{
    NSString* jsCallBack = [NSString stringWithFormat:@"window.HTMLCommandElement.elements[%d].attribute.action();", item.tag];    
    [super writeJavascript:jsCallBack];
}

#pragma mark -
#pragma mark ToolBar

- (void) removeToolBar:(NSArray*)arguments withDict:(NSDictionary*)options
{
	if (!self.toolBar) {
		return;
	}
	
	[self.webView pg_removeSiblingView:self.toolBar withAnimation:NO];
	[self.webView pg_relayout:NO];
	
	self.toolBar = nil;
}

- (void) createToolBar:(NSArray*)arguments withDict:(NSDictionary*)options
{
	if (self.toolBar) {
		return;
	}
	
    CGFloat height   = 45.0f;
    BOOL atTop       = YES;
    UIBarStyle style = UIBarStyleBlackOpaque;

    NSDictionary* toolBarSettings = options;
    if (toolBarSettings) 
	{
        if ([toolBarSettings objectForKey:@"height"])
            height = [[toolBarSettings objectForKey:@"height"] floatValue];
		
        if ([toolBarSettings objectForKey:@"position"]) {
            atTop  = [[toolBarSettings objectForKey:@"position"] isEqualToString:@"top"];
		}
        
        NSString *styleStr = [toolBarSettings objectForKey:@"style"];
        if ([styleStr isEqualToString:@"Default"])
            style = UIBarStyleDefault;
        else if ([styleStr isEqualToString:@"BlackOpaque"])
            style = UIBarStyleBlackOpaque;
        else if ([styleStr isEqualToString:@"BlackTranslucent"])
            style = UIBarStyleBlackTranslucent;
    }

    CGRect toolBarBounds = self.webView.bounds;
	toolBarBounds.size.height = height;
	
    self.toolBar = [[UIToolbar alloc] initWithFrame:toolBarBounds];
    [self.toolBar sizeToFit];
    self.toolBar.hidden                 = NO;
    self.toolBar.multipleTouchEnabled   = NO;
    self.toolBar.autoresizesSubviews    = YES;
    self.toolBar.userInteractionEnabled = YES;
    self.toolBar.barStyle               = style;
	
	self.lastToolBarPosition = atTop?PGLayoutPositionTop:PGLayoutPositionBottom;
	[self.webView pg_addSiblingView:self.toolBar withPosition:self.lastToolBarPosition withAnimation:NO];
}

- (void) toolBarDidSelectItem:(UIBarButtonItem*)item
{
    NSString* jsCallBack = [NSString stringWithFormat:@"window.HTMLCommandElement.elements[%d].attribute.action();", item.tag];
    [super writeJavascript:jsCallBack];
}

- (void) createToolBarItem:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.toolBar) {
        [self createToolBar:nil withDict:nil];
	}

    NSString* name      = [arguments objectAtIndex:0];
    NSString* title     = [arguments objectAtIndex:1];
    NSString* imageName = [arguments objectAtIndex:2];
    BOOL enable         = [[arguments objectAtIndex:3] boolValue];
    int tag             = [[arguments objectAtIndex:4] intValue];
	
    UIBarButtonItem* item = nil;    
    if ([imageName length] > 0) 
	{
        UIBarButtonSystemItem systemItem = -1;
        if ([imageName isEqualToString:@"toolbarButton:Done"])       systemItem = UIBarButtonSystemItemDone;
        if ([imageName isEqualToString:@"toolbarButton:Cancel"])       systemItem = UIBarButtonSystemItemCancel;
        if ([imageName isEqualToString:@"toolbarButton:Edit"])       systemItem = UIBarButtonSystemItemEdit;
        if ([imageName isEqualToString:@"toolbarButton:Save"])       systemItem = UIBarButtonSystemItemSave;
        if ([imageName isEqualToString:@"toolbarButton:Add"])       systemItem = UIBarButtonSystemItemAdd;
        if ([imageName isEqualToString:@"toolbarButton:FlexibleSpace"])       systemItem = UIBarButtonSystemItemFlexibleSpace;
        if ([imageName isEqualToString:@"toolbarButton:FixedSpace"])       systemItem = UIBarButtonSystemItemFixedSpace;
        if ([imageName isEqualToString:@"toolbarButton:Compose"])       systemItem = UIBarButtonSystemItemCompose;
        if ([imageName isEqualToString:@"toolbarButton:Reply"])       systemItem = UIBarButtonSystemItemReply;
        if ([imageName isEqualToString:@"toolbarButton:Action"])       systemItem = UIBarButtonSystemItemAction;
        if ([imageName isEqualToString:@"toolbarButton:Organize"])       systemItem = UIBarButtonSystemItemOrganize;
        if ([imageName isEqualToString:@"toolbarButton:Bookmarks"])       systemItem = UIBarButtonSystemItemBookmarks;
        if ([imageName isEqualToString:@"toolbarButton:Search"])       systemItem = UIBarButtonSystemItemSearch;
        if ([imageName isEqualToString:@"toolbarButton:Refresh"])       systemItem = UIBarButtonSystemItemRefresh;
        if ([imageName isEqualToString:@"toolbarButton:Stop"])       systemItem = UIBarButtonSystemItemStop;
        if ([imageName isEqualToString:@"toolbarButton:Camera"])       systemItem = UIBarButtonSystemItemCamera;
        if ([imageName isEqualToString:@"toolbarButton:Trash"])       systemItem = UIBarButtonSystemItemTrash;
        if ([imageName isEqualToString:@"toolbarButton:Play"])       systemItem = UIBarButtonSystemItemPlay;
        if ([imageName isEqualToString:@"toolbarButton:Pause"])       systemItem = UIBarButtonSystemItemPause;
        if ([imageName isEqualToString:@"toolbarButton:Rewind"])       systemItem = UIBarButtonSystemItemRewind;
        if ([imageName isEqualToString:@"toolbarButton:FastForward"])       systemItem = UIBarButtonSystemItemFastForward;
        if ([imageName isEqualToString:@"toolbarButton:Undo"])       systemItem = UIBarButtonSystemItemUndo;
        if ([imageName isEqualToString:@"toolbarButton:Redo"])       systemItem = UIBarButtonSystemItemRedo;

        if (systemItem != -1) {
            item = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:systemItem target:self action:@selector(toolBarDidSelectItem:)];
			item.style = UIBarButtonItemStyleBordered;
			item.tag = tag;
		}
    }
    
    if (item == nil) 
	{
		if (title && [title length] > 0) {
			NSLog(@"Creating toolbarItem with title");
			item = [[UIBarButtonItem alloc] initWithTitle:title style:UIBarButtonItemStyleBordered 
												   target:self action:@selector(toolBarDidSelectItem:)];
		} else {
			UIImage* image = [UIImage imageNamed:imageName];
			if (!image) {
				NSString* imagePath = [[PhoneGapDelegate class] pathForResource:imageName];
				image = [UIImage imageWithContentsOfFile:imagePath];
			}
			
			NSLog(@"Creating toolbarItem with custom image");
			item = [[UIBarButtonItem alloc] initWithImage:image style:UIBarButtonItemStyleBordered
												   target:self action:@selector(toolBarDidSelectItem:)];
		}
		item.tag = tag;
    }
	
	item.enabled = enable;
	
    BOOL animateItems = YES;
    if ([options objectForKey:@"animate"]) {
        animateItems = [(NSString*)[options objectForKey:@"animate"] boolValue];
	}
	
    NSMutableArray* items = [[self.toolBar items] mutableCopy];
	if (!items) {
		items = [[NSMutableArray alloc] initWithCapacity:1];
	}
	[items addObject:item];

    [self.toolBar setItems:items animated:YES];
	[items release];
	
	[self.toolBarItems setObject:item forKey:name];
}

/**
 * Remove an existing toolbar item
 * @param arguments Parameters used to identify the toolbar item to update
 *  -# \c name internal name used to represent this item when it was created
 */
- (void) removeToolBarItem:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.toolBar) {
        [self createToolBar:nil withDict:nil];
	}
	
    NSString* name = [arguments objectAtIndex:0];
    UIBarButtonItem* item = [self.toolBarItems objectForKey:name];
	
	NSMutableArray* items = [self.toolBar.items mutableCopy];
	NSUInteger index = [items indexOfObject:item];
	if (index != NSNotFound)
	{
		[items removeObjectAtIndex:index];
		[self.toolBar setItems:items animated:YES];
	}
	
	[items release];
}

- (void) enableToolBarItem:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.toolBar) {
        [self createToolBar:nil withDict:nil];
	}
	
    NSString* name      = [arguments objectAtIndex:0];
    BOOL enable         = [[arguments objectAtIndex:1] boolValue];
	
    UIBarButtonItem* item = [self.toolBarItems objectForKey:name];
	
	NSUInteger index = [self.toolBar.items indexOfObject:item];
	if (index != NSNotFound)
	{
		item.enabled = enable;
	}
}	

/**
 * Show the toolbar after its been created.
 */
- (void) showToolBar:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.toolBar) {
        [self createToolBar:nil withDict:nil];
	}
    
	if (![self.webView pg_hasSiblingView:self.toolBar]) {
		[self.webView pg_addSiblingView:self.toolBar withPosition:self.lastToolBarPosition withAnimation:NO];
	}
}

/**
 * Hide the toolbar
 */
- (void) hideToolBar:(NSArray*)arguments withDict:(NSDictionary*)options
{
    if (!self.toolBar) {
        return;
	}
    
	if ([self.webView pg_hasSiblingView:self.toolBar]) {
		[self.webView pg_removeSiblingView:self.toolBar withAnimation:NO];
		[self.webView pg_relayout:NO];
	}
}

@end
